### DISCUSSION TITLE
Variable Invocation

### DISCUSSION THREAD

**Discussion by Arpita-Jaiswal** [fastn-stack/fastn#424] [General]
Currently, there's no way to determine after the `p1 parser` whether the invocation is of component or variable.

# Problem 1: Variable vs Component Invocation

How to determine whether the invocation is of component or variable?

Consider this variable Invocation
```ftd
-- age: 40

-- abrar:
name: AbrarK
age: 10
```

for definition like

```ftd
-- integer age: 30

-- record person:
string name:
integer age:

-- person abrar:
name: Abrar Khan
age: 5
```

Also, consider the component invocation

```ftd
-- abrar:
name: AbrarK
age: 10

-- ftd.column abrar:
string name:
integer age:

-- ftd.text: $name
-- ftd.integer: $age

-- end: abrar 
```

Both look the same, and we need to check the definition to know whether this is a component or variable invocation. So, how do we resolve this syntactically?

So if for any variable invocation we will use `$` right before the variable name. Also, now we can invoke a variable only if the variable is mutable and this should be ensured in the definition itself. You can define a variable mutable by adding `$` as the prefix in the variable name.

Consider this
```ftd
;; Mutable Variable Definition
-- person $abrar:
name: Abrar Khan
age: 5

;; Variable Invocation
-- $abrar:
name: AbrarK
age: 10

;; Component Invocation
-- abrar:
name: AbrarK
age: 10
```

It is now distinct.

# Problem 2:  Variable update vs insert

How can we determine whether the variable invocation is to update the variable or to insert the data in the list type variable?

According to our current syntax, if the variable is of list type then its invocation would append the item in the list. On the other hand, if the variable is a simple type variable, then its invocation would update the variable value.

```ftd
-- string list $names:

;; This invocation results in appending `Arpita` in names
--  $names: Arpita



-- string $name: Abrar

;; This invocation results in variable updation
-- $name: AmitU
```

The proposed syntax says that the variable invocation is always variable updation for any kind/type of variable.

Now how to initialise the list type variable with some value?

This is enclosed in [anonymous list discussion](https://github.com/FifthTry/ftd/discussions/385).

**Comment by sharmashobhit** [comment]
```ftd

-- mut string abrar: Abrar Khan

-- $abrar: AbrarK

-- component abrar:
string name:

-- ftd.row: $name

-- end: ftd.row
-- end: abrar

-- abrar:
name: $abrar

```

**Comment by Arpita-Jaiswal** [comment]
```ftd
-- component foo:
string $name:

-- ftd.text: Click Here
$on-click$: $name = Arpita

-- end: foo

-- string $a-name: Abrark

-- foo:
$name: $a-name

-- ftd.text: $a-name
```

```ftd
-- component foo:
string $name: Abrar

-- ftd.text: Click Here
$on-click$: $name = Arpita

-- end: foo

-- string $a-name: Abrark

-- foo:

-- foo:
$name: $a-name

-- foo:
$name: Abrar

-- ftd.text: $a-name
```

**Comment by Arpita-Jaiswal** [comment]
```ftd
-- record person:
string name:
string gender:

-- string $name: Name0

-- person $n1: 
name: $name ;; clone of variable name
gender: f

-- person $a1: $n1

-- $a1.name: Name1
-- $n1.name: Name2
```

## Mutable fields in the record not allowed

```ftd
-- record person:
string $name:
string gender:
```

This throws error

## Assigning non-mutable reference to mutable is an error 

```ftd
-- person n1: 
name: $name
gender: f

-- person $a1: $n1
```
The above throws an error, but we can pass the clone.

```ftd
-- person n1: 
name: $name
gender: f

-- person $a1: *$n1
```

## TODO:

Later on, we'll support passing the mutable reference in the field value 
```ftd
-- person $n1: 
$name: $name ;; mutable reference of variable name
gender: f
```
